src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java

```diff
@@ -595,6 +603,104 @@ public void respawn()
+	 * @return {@link PlayerInteractEvent} that was fired
+	 */
+	public @NotNull PlayerInteractEvent simulateUseItemOn(@NotNull Location location, @NotNull BlockFace clickedFace, @Nullable EquipmentSlot hand) {
+		Preconditions.checkNotNull(location, "Location cannot be null");
+		Preconditions.checkArgument(hand == EquipmentSlot.HAND || hand == EquipmentSlot.OFF_HAND, "Must be HAND or OFF_HAND");
+
+		ItemStack itemInHand = hand == EquipmentSlot.HAND ? getItemInHand() : getInventory().getItemInOffHand();
+
+		boolean cancelled = false;
+		BlockMock block;
+		if (location.getBlock() instanceof BlockMock blockMock) {
+			block = blockMock;
+		} else {
+			block = new BlockMock(location);
+		}
+
+		if (getGameMode() == GameMode.SPECTATOR) {
+			cancelled = !(block.getState() instanceof Container);
+		}
+
+		// TODO Implement Cooldown
+//		if (hasCooldown(itemInHand.getType())) {
+//			cancelled = true;
+//		}
+
+		ItemStack eventItemStack = itemInHand;
+		if (eventItemStack.getType() == Material.AIR || eventItemStack.getAmount() == 0) {
+			eventItemStack = null;
+		}
+
+		PlayerInteractEvent event = new PlayerInteractEvent(this, Action.RIGHT_CLICK_BLOCK, eventItemStack, block, clickedFace, hand);
+		if (cancelled) {
+			event.setUseInteractedBlock(Event.Result.DENY);
+		}
+		Bukkit.getServer().getPluginManager().callEvent(event);
+
+		boolean interactResult = event.useItemInHand() == Event.Result.DENY;
+
+		if (event.useInteractedBlock() == Event.Result.DENY) {
+			return event;
+		} else if (getGameMode() == GameMode.SPECTATOR) {
+			if (block.getState() instanceof ContainerMock containerMock) {
+				openInventory(containerMock.getInventory());
+			}
+		} else {
+			boolean isItemInHand = !getItemInHand().isEmpty() || !getInventory().getItemInOffHand().isEmpty();
+			boolean shouldCancelUse = isSneaking() && isItemInHand;
+			ItemStack itemStackCopy = itemInHand.clone();
+
+			if (!shouldCancelUse) {
+				if (block.getState() instanceof BlockStateMock blockStateMock) {
+					EnumInteractionResult result = blockStateMock.simulateUse(this, hand);
+					block.setState(blockStateMock);
+					block.setBlockData(blockStateMock.getBlockData());
+
+					if (result.consumesAction()) {
+						// TODO implement CriterionTriggers
+						return event;
+					}
+				} else {
+					throw new UnimplementedOperationException();
+				}
+			}
+
+			if (!itemStackCopy.isEmpty() && !interactResult) {
+				EnumInteractionResult itemInteractionResult;
+
+				ItemStackMock itemStackMock;
+				if (itemStackCopy instanceof ItemStackMock) {
+					itemStackMock = (ItemStackMock) itemStackCopy;
+				} else {
+					itemStackMock = new ItemStackMock(itemStackCopy);
+				}
+
+				if (getGameMode() == GameMode.CREATIVE) {
+					int count = itemStackMock.getAmount();
+					itemInteractionResult = itemStackMock.simulateUse(this, location, hand);
+					itemStackCopy.setAmount(count);
+				} else {
+					itemInteractionResult = itemStackMock.simulateUse(this, location, hand);
+				}
```

Comment:
Looked into the sourcecode for paper, and you're not replicating the behavior here, these operations should not be run on a copy on the original itemstack, but actually on the original itemstack itself. So to say, you're instansiating a local variable itemStackCopy, and then you are modifying that instance: It's doing nothing.

(Note that I'm ignoring the fact that running this throws an UnimplementedOperationException anyways)